### ISSUE TITLE
Scripting API: nil object error when comparing different types

### ISSUE THREAD

**Issue by lampysprites** [aseprite/aseprite#3218]
Consider this example:
```lua
print(app.activeSprite.layers[1] == app.activeSprite)
```
Intuitively I'd expect the expression to simply return false, same as `"some string" == 33` for instance. Currently it shows `using a nil 'doc::Layer' object` error. That is a bit misleading: although the object can be nil there, usually it means that the type cast failed internally. 

Posting here for the sake of documentation, functionally it's a minor issue. On the script side it can be wrapped in a pcall:
```lua
local function docobjEquals(a, b)
    local ok, eq = pcall(function () return a == b end)
    return ok and eq
end
```

### Aseprite and System version

* Aseprite version: 1.3-beta14-x64 steam
* System: Windows 10 x64

**Comment by behreajj** [comment]
Hi,

Thanks so much for pointing this out.


I hope no one minds, but I want to leave a note here, mainly for myself if no one else.
The message comes from here:
https://github.com/aseprite/aseprite/blob/b5414c5d952f152cdf336f048d88469a91b75d57/src/app/script/docobj.h#L44

**Comment by lampysprites** [comment]
It sure takes time to figure out the way around the codebase ( x _ x  ).

It's `get_docobj` in every type metatable's `__eq` that assumes the type of the other operand to be the same, finds no object with given type and ID (bc it's a different type), and throws because of that. e.g.

https://github.com/aseprite/aseprite/blob/b5414c5d952f152cdf336f048d88469a91b75d57/src/app/script/sprite_class.cpp#L140

https://github.com/aseprite/aseprite/blob/b5414c5d952f152cdf336f048d88469a91b75d57/src/app/script/sprite_class.cpp#L812

One possible way is using `may_get_docobj` instead, that does not throw, and do null checks manually, returning false when it makes sense (tho what about two nulls?). 

Or perhaps it'll be easier first make preliminary checks with [lua_isuserdata](https://www.lua.org/manual/5.4/manual.html#lua_isuserdata) and [luaL_checkudata](https://www.lua.org/manual/5.4/manual.html#luaL_checkudata) before trying to retrieve docobj's. For that, the types are defined by the metatable, just like normally in lua. There's some macro stuff in the end of the file to define them, it seems?

**Comment by behreajj** [comment]
Yeah... I thought it was mutually understood that we were talking about the equality operators of sprite and layer. That's how I found my way to the method I linked above. Though, I should mention that the possible return types of a layer's parent are why this is a recurring issue for me.

I realize it varies with how operator overloading is handled across languages, but I was under the impression that the type of the right operand that matters first - or more - because the equality method would be called by the left operand?

For an analogous problem, imagine I have a 2D vector class and want to implement leftVector * rightScalar and leftScalar * rightVector. When vector is on the left, I could overload the multiply operator in the vector class. It's less clear for the second case. If the left operand is nil, I would hope that neither a layer's nor a sprite's equality method would even be called.

Thanks for the other links.

**Comment by behreajj** [comment]
Additional clarification: Maybe I'm assuming that the responsibility for how to handle null is left to each class definition's implementation of the operator.

**Comment by lampysprites** [comment]
Apparently, you're correct that for equality it isn't... it's only actually
called for two different tables/userdatas
https://www.lua.org/manual/5.4/manual.html#2.4
So that's why we don't have problems for explicit null check huh.

**Comment by behreajj** [comment]
Yeah, I appreciate the subtle hints to RTFM. I do this in my free time, for free. It's not my job to research and redesign equality for the whole API according to the best standards and practices of C++. I just want it to stop getting in the way. I realize it's the weekend; no one has to answer until Monday.

**Comment by lampysprites** [comment]
I'm so sorry that it came off like an rtfm or else condescending, really don't mean it that way. Considering the initial question was about checking with C++ and not Lua, I read that as if you wanted to solve the issue by patching the API.
I realize I failed at helping already, but fwiw, on script side type check could be `type(parent) == "userdata" and getmetatable(parent).__name == "doc::Sprite"` - that should stop the error getting in the way?

**Comment by behreajj** [comment]
I'm sorry for losing my temper and for being unclear about where and in what language I was working on.

I was able to patch something together in C++ with `lua_isnoneornil(L, 2)` and `luaL_testudata(L, 2, "Layer");` I do appreciate your help - sincerely, this time - once I reread your responses they lead me to what I was looking for.

(From what I've tested so far, it doesn't solve the case where, in Lua, someone can call `sprite:close()` then call a method from a layer contained by that sprite. -- I'm just leaving that note in case that's an indicator for others looking to solve broader issues.)
